File: pyjsps.js

function PyPacket(label, args) {
  this.label = label;
  if(args != undefined) this.args = args;
  else this.args = [];
  this.parse = function() {
    var mapped = this.label;
    for(var i = 0; i < this.args.length; i++)
      mapped += PySocket.splitter + this.args[i];
    return mapped;
  }
  this.response = undefined;
  this.isPyPacket = true;
}

class PySocket {

  static PREFIX = "[PySocket]: ";

  static splitter = '\n';

  constructor(address, port, secured) {
    this.address = address;
    this.port = port;
    this.debug = true;
    if(secured != undefined)
    this.ssl = true;
  }

  static package = function(recv) {
    var mapped = recv.split(PySocket.splitter);
    if(mapped.length <= 0) return undefined;
    else {
      var label = mapped[0];
      var args = [];
      for(var i = 1; i < mapped.length; i++)
        args[i - 1] = mapped[i];
      return new PyPacket(label, args);
    }
  }

  log(info) {
    if(this.debug)
    console.log(PySocket.PREFIX + info);
  }

  send(packet, recv) {
    if(!packet.isPyPacket) {
      this.log("Script tried to send wrong Packet-Type: " + (typeof packet));
      return undefined;
    }
    var server = this.address + ':' + this.port;
    var socket = new WebSocket((this.ssl ? "wss://" : "ws://") + server);
    this.log("Sending to \"" + server  + "\": " + packet.label);
    socket.onopen = function(event) { socket.send(packet.parse()); }
    if(recv) socket.onmessage = function(event) { recv(PySocket.package(event.data)); }
  }

}

File: pyjsps.py

import asyncio
import websockets
import ssl, os
import pathlib

HOST = "127.0.0.1";

info_splitter = '\n'

class JsPacket:

    def __init__(self, label, args=[], sender=None):
        self.label = label
        self.args = args
        self.sender = sender

    def parse(self):
        mapped = self.label
        for arg in self.args:
            mapped += info_splitter + arg
        return mapped

class JsSocket:

    def __init__(self, port, handler, cert=None):
        self.ssl_context = None
        self.port = port
        if cert != None:
            ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
            ssl_context.load_cert_chain(pathlib.Path(os.getcwd() + f"/{cert}/cert.pem"), keyfile=(os.getcwd() + f"/{cert}/key.pem"))
        self.handler = handler
        self.prefix = "[JsSocket]: "

    async def handshake(self, client, address):
        data = await client.recv()
        mapped = data.split(info_splitter)
        label = mapped[0]
        args = []
        for i in range(len(mapped) - 1): args.append(mapped[i + 1])
        packet = self.handler(JsPacket(label, args))
        await client.send(packet.parse())

    def listen_forever(self, loop=None):
        if loop is None: loop = asyncio.get_event_loop()
        else: asyncio.set_event_loop(loop)
        if self.ssl_context is not None:
            self.server = websockets.serve(self.handshake, HOST, self.port, ssl=self.ssl_context)
        else:
            self.server = websockets.serve(self.handshake, HOST, self.port)
        loop.run_until_complete(self.server)
        loop.run_forever()
